File: Samples/Verbum/how-to-implement-the-insert-document-field-dialog-winforms-version-t223036-14.2.6-/VB/WindowsFormsApplication1/InsertFieldModules/InsertFieldRichEditHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraBars.Ribbon
Imports DevExpress.XtraBars
Imports System.Drawing
Imports System.Windows.Forms

Namespace WindowsFormsApplication1.InsertFieldModules

    Public Module InsertFieldRichEditHelper
        Private privateRichControl As RichEditControl
        Public Property RichControl() As RichEditControl
            Get
                Return privateRichControl
            End Get
            Set(ByVal value As RichEditControl)
                privateRichControl = value
            End Set
        End Property
        Private privateInsertDialog As InsertFieldDialog
        Public Property InsertDialog() As InsertFieldDialog
            Get
                Return privateInsertDialog
            End Get
            Set(ByVal value As InsertFieldDialog)
                privateInsertDialog = value
            End Set
        End Property

        Private _variableTypes As New System.Collections.Generic.SortedDictionary(Of String, UserControl)()
        Public ReadOnly Property VariableTypes() As System.Collections.Generic.SortedDictionary(Of String, UserControl)
            Get
                Return _variableTypes
            End Get
        End Property
        Private privateCurrentVariableType As String
        Public Property CurrentVariableType() As String
            Get
                Return privateCurrentVariableType
            End Get
            Set(ByVal value As String)
                privateCurrentVariableType = value
            End Set
        End Property


        <System.Runtime.CompilerServices.Extension()> _
        Public Sub RegisterInsertFieldDialogCommand(ByVal richEdit As RichEditControl, ByVal pageGroup As RibbonPageGroup, ByVal commandImage As Image)
            Dim buttonInsertFieldDialog As New BarButtonItem(pageGroup.Ribbon.Manager, "Insert Field")
            buttonInsertFieldDialog.LargeGlyph = If(commandImage IsNot Nothing, commandImage, Image.FromFile("InsertFieldModules\\insertFieldIcon.png"))
            AddHandler buttonInsertFieldDialog.ItemClick, AddressOf buttonInsertFieldDialog_ItemClick
            pageGroup.ItemLinks.Add(buttonInsertFieldDialog)

            RichControl = richEdit
        End Sub


        Private Sub CreateVariablesTypes()
            _variableTypes.Add("DOCVARIABLE", New ucDOCVARIABLE())
            _variableTypes.Add("PAGE", New ucPAGENUMBER("PAGE"))
            _variableTypes.Add("NUMPAGES", New ucPAGENUMBER("NUMPAGES"))
            _variableTypes.Add("CREATEDATE", New ucCREATEDATE())
            _variableTypes.Add("DATE", New ucCREATEDATE())
            _variableTypes.Add("TIME", New ucCREATEDATE())
            _variableTypes.Add("IF", New ucIFFIELD())

        End Sub

        Private Sub buttonInsertFieldDialog_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            If VariableTypes.Keys.Count = 0 Then
                CreateVariablesTypes()
            End If
            If InsertDialog Is Nothing Then
                InsertDialog = New InsertFieldDialog(RichControl)
            End If
            InsertDialog.ShowDialog()
        End Sub

        Public Const EmptyFieldCode As String = "EMPTYFIELD"
    End Module

    ' Additional service classes
    Public Interface IFieldParametersDialog
        Function GenerateFieldParametersString() As String
    End Interface

    Public Class FieldAttributeItem
        Private privateDisplayName As String
        Public Property DisplayName() As String
            Get
                Return privateDisplayName
            End Get
            Set(ByVal value As String)
                privateDisplayName = value
            End Set
        End Property
        Private privateAttributeValue As String
        Public Property AttributeValue() As String
            Get
                Return privateAttributeValue
            End Get
            Set(ByVal value As String)
                privateAttributeValue = value
            End Set
        End Property
    End Class
End Namespace

File: Samples/Verbum/how-to-implement-the-insert-document-field-dialog-winforms-version-t223036-14.2.6-/CS/WindowsFormsApplication1/InsertFieldModules/InsertFieldRichEditHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraRichEdit;
using DevExpress.XtraBars.Ribbon;
using DevExpress.XtraBars;
using System.Drawing;
using System.Windows.Forms;

namespace WindowsFormsApplication1.InsertFieldModules {

    public static class InsertFieldRichEditHelper {
        public static RichEditControl RichControl { get; set; }
        public static InsertFieldDialog InsertDialog { get; set; }

        private static System.Collections.Generic.SortedDictionary<string, UserControl> _variableTypes = new System.Collections.Generic.SortedDictionary<string, UserControl>();
        public static System.Collections.Generic.SortedDictionary<string, UserControl> VariableTypes { get { return _variableTypes; } }
        public static string CurrentVariableType { get; set; }


        public static void RegisterInsertFieldDialogCommand(this RichEditControl richEdit, RibbonPageGroup pageGroup, Image commandImage) {
            BarButtonItem buttonInsertFieldDialog = new BarButtonItem(pageGroup.Ribbon.Manager, "Insert Field");
            buttonInsertFieldDialog.LargeGlyph = commandImage != null ? commandImage : Image.FromFile("InsertFieldModules\\insertFieldIcon.png");
            buttonInsertFieldDialog.ItemClick += new ItemClickEventHandler(buttonInsertFieldDialog_ItemClick);
            pageGroup.ItemLinks.Add(buttonInsertFieldDialog);

            RichControl = richEdit;
        }


        private static void CreateVariablesTypes() {
            _variableTypes.Add("DOCVARIABLE", new ucDOCVARIABLE());
            _variableTypes.Add("PAGE", new ucPAGENUMBER("PAGE"));
            _variableTypes.Add("NUMPAGES", new ucPAGENUMBER("NUMPAGES"));
            _variableTypes.Add("CREATEDATE", new ucCREATEDATE());
            _variableTypes.Add("DATE", new ucCREATEDATE());
            _variableTypes.Add("TIME", new ucCREATEDATE());
            _variableTypes.Add("IF", new ucIFFIELD());
            
        }

        static void buttonInsertFieldDialog_ItemClick(object sender, ItemClickEventArgs e) {
            if(VariableTypes.Keys.Count == 0) {
                CreateVariablesTypes();    
            }
            if(InsertDialog == null) {
                InsertDialog = new InsertFieldDialog(RichControl);    
            }
            InsertDialog.ShowDialog();
        }

        public const string EmptyFieldCode = "EMPTYFIELD";
    }

    // Additional service classes
    public interface IFieldParametersDialog {
        string GenerateFieldParametersString();
    }

    public class FieldAttributeItem {
        public string DisplayName { get; set; }
        public string AttributeValue { get; set; }    
    }
}
